### PR TITLE
Challenges 1-4 in Java

### PR DESCRIPTION
**Pull request by zmiller91**

### CODE REVIEW

File: challenge_4/java/zmiller91/src/zmiller91/InvertBinaryTree.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+public class InvertBinaryTree {
+    
+    /**
+     * Post order tree traversal to invert the tree from the bottom up.
+     * 
+     * @param root - binary tree
+     */
+    public static void invert(Node root) {
+        
+        // Stop at the n-1 row
```
Comment (moonpatroller) [inline challenge_4/java/zmiller91/src/zmiller91/InvertBinaryTree.java]: You probably want to invert if it's got at least one child.  The child could be the root of a full tree.

File: challenge_4/java/zmiller91/src/zmiller91/Queue.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package zmiller91;
+
+import java.util.LinkedList;
+
+/**
+ * Generic Queue data structure. It's a wrapper to a linked list.
+ * 
+ * @author zmiller
+ */
+public class Queue<T> {
```
Comment (moonpatroller) [inline challenge_4/java/zmiller91/src/zmiller91/Queue.java]: Why write this wrapper when this already exists in the Java standrd libs?

Comment (zmiller91) [inline challenge_4/java/zmiller91/src/zmiller91/Queue.java]: There's no reason.

File: challenge_4/java/zmiller91/src/zmiller91/InvertBinaryTree.java (hunk 1 above)
Comment (zmiller91) [inline challenge_4/java/zmiller91/src/zmiller91/InvertBinaryTree.java]: Nice catch!

Comment (zmiller91) [inline challenge_4/java/zmiller91/src/zmiller91/InvertBinaryTree.java]: After further testing, the original logic works as intended. Unless you're seeing something I'm not, this works for: 

    * Single node trees
    * Balanced, complete binary trees
    * Balanced, incomplete binary trees
    * Unbalanced binary trees
    * Trees with at most 1 node

I'm curious as to what your thoughts are, though.

**Comment by jarydremillard** [comment]
- [ ] Code review
- [ ] Correct path
- [ ] Code tested
- [ ] Check for conflicts
- [ ] **Merge PR**

**Comment by jarydremillard** [comment]
bypass

**Pull request merged**